Makefile

```diff
@@ -803,56 +577,491 @@ $(expunge): compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma \
+  $(eval $(call COMPILE_C_FILE,runtime/$(runtime_OBJECT_TYPE),runtime/%)))
+
+runtime/$(UNIX_OR_WIN32)_non_shared.%.$(O): \
+  OC_CPPFLAGS += -DBUILDING_LIBCAMLRUNS
+
+$(eval $(call COMPILE_C_FILE,\
+  runtime/$(UNIX_OR_WIN32)_non_shared.%, runtime/$(UNIX_OR_WIN32)))
+
+$(foreach runtime_OBJECT_TYPE,$(subst %,,$(runtime_OBJECT_TYPES)), \
+  $(eval \
+    runtime/dynlink$(runtime_OBJECT_TYPE).$(O): $(ROOTDIR)/Makefile.config))
+
+## Compilation of runtime assembly files
+
+ASPP_ERROR = \
+  { echo "If your assembler produced syntax errors, it is probably";\
+          echo "unhappy with the preprocessor. Check your assembler, or";\
+          echo "try producing $*.o by hand.";\
+          exit 2; }
+runtime/%.o: runtime/%.S
+	$(ASPP) $(OC_ASPPFLAGS) -o $@ $< || $(ASPP_ERROR)
+
+runtime/%.d.o: runtime/%.S
+	$(ASPP) $(OC_ASPPFLAGS) $(OC_DEBUG_CPPFLAGS) -o $@ $< || $(ASPP_ERROR)
+
+runtime/%.i.o: runtime/%.S
+	$(ASPP) $(OC_ASPPFLAGS) $(OC_INSTR_CPPFLAGS) -o $@ $< || $(ASPP_ERROR)
+
+runtime/%_libasmrunpic.o: runtime/%.S
+	$(ASPP) $(OC_ASPPFLAGS) $(SHAREDLIB_CFLAGS) -o $@ $<
+
+runtime/domain_state64.inc: \
+  runtime/gen_domain_state64_inc.awk runtime/caml/domain_state.tbl
+	$(AWK) -f $^ > $@
+
+runtime/domain_state32.inc: \
+  runtime/gen_domain_state32_inc.awk runtime/caml/domain_state.tbl
+	$(AWK) -f $^ > $@
+
+runtime/amd64nt.obj: runtime/amd64nt.asm runtime/domain_state64.inc
+	$(ASM)$@ $<
+
+runtime/i386nt.obj: runtime/i386nt.asm runtime/domain_state32.inc
+	$(ASM)$@ $<
+
+runtime/amd64nt.d.obj: runtime/amd64nt.asm runtime/domain_state64.inc
+	$(ASM)$@ $(ocamlrund_CPPFLAGS) $<
+
+runtime/i386nt.d.obj: runtime/i386nt.asm runtime/domain_state32.inc
+	$(ASM)$@ $(ocmalrund_CPPFLAGS) $<
+
+runtime/amd64nt.i.obj: runtime/amd64nt.asm runtime/domain_state64.inc
+	$(ASM)$@ $(ocamlruni_CPPFLAGS) $<
+
+runtime/i386nt.i.obj: runtime/i386nt.asm runtime/domain_state32.inc
+	$(ASM)$@ $(ocamlruni_CPPFLAGS) $<
+
+runtime/%_libasmrunpic.obj: runtime/%.asm
+	$(ASM)$@ $<
+
+## Runtime dependencies
+
+runtime_DEP_FILES := $(addsuffix .b, \
+  $(basename $(runtime_BYTECODE_C_SOURCES) runtime/instrtrace))
+ifneq "$(NATIVE_COMPILER)" "false"
+runtime_DEP_FILES += $(addsuffix .n, $(basename $(runtime_NATIVE_C_SOURCES)))
+endif
+runtime_DEP_FILES += $(addsuffix d, $(runtime_DEP_FILES)) \
+             $(addsuffix i, $(runtime_DEP_FILES)) \
+             $(addsuffix pic, $(runtime_DEP_FILES))
+runtime_DEP_FILES := $(addsuffix .$(D), $(runtime_DEP_FILES))
+
+ifeq "$(COMPUTE_DEPS)" "true"
+include $(addprefix $(DEPDIR)/, $(runtime_DEP_FILES))
+endif
+
+.PHONY: runtime
+runtime: stdlib/libcamlrun.$(A)
+
+ifeq "$(BOOTSTRAPPING_FLEXDLL)" "true"
+runtime: $(addprefix stdlib/flexdll/, $(FLEXDLL_OBJECTS))
+stdlib/flexdll/flexdll%.$(O): $(FLEXDLL_SOURCES)/flexdll%.$(O) | stdlib/flexdll
+	cp $< $@
+stdlib/flexdll:
+	$(MKDIR) $@
+endif
+
+.PHONY: makeruntime
+makeruntime:
+	$(MAKE) $(BOOT_FLEXLINK_CMD) runtime-all
+stdlib/libcamlrun.$(A): runtime/libcamlrun.$(A)
```

Comment:
Unfortunately, as with `stdlib/libasmrun.$(A)` below, the dependency needs to be `runtime-all`, or you can't do `make world.opt ; make clean ; make opt.opt` (i.e. this step implicitly updates `runtime/ocamlrun` which has been lost with this change).